### PR TITLE
feat: Select component (Iteration 1)

### PR DESCRIPTION
**Pull request by geido**
### SUMMARY

This PR introduces a new Select component based on the Antdesign Select. The goal of the refactoring is to eliminate all the various instances of selects throughout the codebase and merge them into one, while getting rid of the react-select entirely.

###  PROPS
At the moment only _some_ of the Antd props have been exposed in the interface on top of the custom props. This is to make sure that the Select component is used with the intended behaviors. However, as the implementation of the new component goes, we might realize that other default props should be exposed. 

| Prop Name         | Origin | Type                                                                             | Description                                                                                                                                                                                                                                             | Prop exposed | Default behavior (prop not exposed) |
| ----------------- | ------ | -------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------ | ----------------------------------- |
| mode              | Antd   | multiple OR tags                                                                 | Set mode of Select                                                                                                                                                                                                                                      | x            |                                     |
| truncateOptions   | Custom | true (default) OR false                                                          | When true it truncates the options with a long text and shows a tooltip. When false, it wraps options with a long text to a new line.                                                                                                                   |              | x                                   |
| maxTagCount       | Antd   | number OR responsive                                                             | Max tag count to show. responsive will cost render performance                                                                                                                                                                                          |              | x                                   |
| showSearch        | Antd   | true OR false (default)                                                          | Whether show search input in single mode                                                                                                                                                                                                                | x            |                                     |
| allowNewOptions   | Custom | true OR false                                                                    | It allows to create a new option when no options match the search. This is true by default when the property mode is set to tags OR multiple. Otherwise it's false by default.                                                                          | x            |                                     |
| options           | Antd   | { label, value }\[\] OR Promise => { label, value }\[\]                          | Select options. It can accept an array or a promise that resolves an array or rejects                                                                                                                                                                   | x            |                                     |
| loading           | Antd   | true OR false (default)                                                          | Indicate loading state                                                                                                                                                                                                                                  | x            |                                     |
| onChange          | Antd   | function(value, option:Option OR Array<Option>)                                  | Called when select an option or input value change                                                                                                                                                                                                      | x            |                                     |
| placeholder       | Antd   | ReactNode                                                                        | Placeholder of select                                                                                                                                                                                                                                   | x            |                                     |
| value             | Antd   | string OR string\[\]<br>number OR number\[\]<br>LabeledValue OR LabeledValue\[\] | Current selected option                                                                                                                                                                                                                                 | x            |                                     |
| tokenSeparators   | Antd   | string\[\]                                                                       | Separator used to tokenize on tag and multiple mode                                                                                                                                                                                                     |              | x                                   |
| autoFocus         | Antd   | true OR false (default)                                                          | Get focus by default                                                                                                                                                                                                                                    | x            |                                     |
| allowClear        | Antd   | true OR false (default)                                                          | Show clear button                                                                                                                                                                                                                                       | x            |                                     |
| disabled          | Antd   | true OR false (default)                                                          | Whether disabled select                                                                                                                                                                                                                                 | x            |                                     |
| filterOption      | Antd   | true (default) OR false OR function(inputValue, option)                          | If true, filter options by input, if function, filter options against it. The function will receive two arguments, inputValue and option, if the function returns true, the option will be included in the filtered set; Otherwise, it will be excluded | x            |                                     |
| aria-label        | Antd   | string                                                                           | Aria label (for assistive tech)<br><br>                                                                                                                                                                                                                 | x            |                                     |
| getPopupContainer | Antd   | function(triggerNode)<br><br>Default:<br>() => document.body                     | Parent Node which the selector should be rendered to. Default to body. When position issues happen, try to modify it into scrollable content and position it relative.                                                                                  |              | x                                   |
| name              | Custom | string                                                                           | Used for the test cases                                                                                                                                                                                                                                 | x            |                                     |
| notFoundContent   | Antd   | ReactNode                                                                        | Specify content to show when no result matches                                                                                                                                                                                                          | x            |                                     |
| header            | Custom | ReactNode                                                                        | Header of the Select                                                                                                                                                                                                                                    | x            |                                     |
| paginatedFetch    | Custom | true OR false (default)                                                          | Makes the async select paginated                                                                                                                                                                                                                        | x            |

### CUSTOM FEATURES
Few abilities have been added on top of the default Antd Select component, as follows:

- The creation of custom options for single mode
- Options as an array or as a Promise to fetch data from the server (more information below in Async Select section)
- Header as a prop
- Sorting of the selected options at the top of the dropdown
- Automatic scroll at the top of the dropdown 

### ASYNC SELECT
The Select component now has the ability to fetch data. The `options` property can accept a `Promise`. The Select component will do this for you:

- Runs the fetch when the component mounts to load the initial options
- It will check if the searched value already exists in the options and won't run the fetch in that case
- It allows pagination when the property `paginatedFetch` is set and will automatically run the fetch by passing the current page as the user scrolls through the options
- It will handle and print errors

### WHAT'S MISSING (Upcoming in Iteration 2)

- Paginate few px before the bottom of the scroll for better UX
- Check/Discuss whether we need the onPaste and on closeOnScrollOut props
- Remove the usage of any from the implementation
- When allowNewOptions is true, showSearch should be true as well in Storybook
- Our interface should only expose single and multiple modes
- Test cases

This first iteration should be merged as it does not affect any existing functionality. @michael-s-molina  will take over the "WHAT'S MISSING" and any other change request in a separate PR as soon as this is merged (Iteration 2).

### CODE REVIEW

**Comment by michael-s-molina** [comment]
@geido First thoughts:

- You can rename `AntdSelect` and `AntSelect.stories` to `Select` and `Select.stories`  using the `index` to resolve any name conflicts.
- I think it would be nice to simulate a service API for the storybook. We can support search, pagination and also throw errors to test the error handling.
- You can use React `useReducer` to manipulate multiple state values if the number of states increases.
- Replace `handleOnDeselect = (option: any)` with `handleOnDeselect = (value: string | number | AntdLabeledValue)`
- Replace `handleOnSelect = (option: any)` with `handleOnSelect = (value: string | number | AntdLabeledValue)`
- It would be nice to add one Select on each corner of the viewport on the storybook page. Then we can test the drop-down behavior for each corner and when scrolling.

**Comment by geido** [comment]
> @geido First thoughts:
> 
> * You can rename `AntdSelect` and `AntSelect.stories` to `Select` and `Select.stories`  using the `index` to resolve any name conflicts.

This is just to simplify the review and to avoid that the new implementation would appear as changes made on the original file. Once this PR is merged, we can just rename it.

> * I think it would be nice to simulate a service API for the storybook. We can support search, pagination, and also throw errors to test the error handling.

I have enhanced the storybook to provide an Async Select dedicated story. However, I am looking to do more improvements there as soon as I am done with the rest of the implementation.

> * You can use React `useReducer` to manipulate multiple state values if the number of states increases.

I am definitely going to if the number of changes to the state should increase.

> * Replace `handleOnDeselect = (option: any)` with `handleOnDeselect = (value: string | number | AntdLabeledValue)`
> * Replace `handleOnSelect = (option: any)` with `handleOnSelect = (value: string | number | AntdLabeledValue)`

I tried, however, typescript still complaints. I am avoiding dedicating too much time to this and leaving it for when the main implementation is completed.

> * It would be nice to add one Select on each corner of the viewport on the storybook page. Then we can test the drop-down behavior for each corner and when scrolling.

This is now done.

Thank you!

File: superset-frontend/src/components/Select/AntdSelect.tsx [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+export enum ESelectTypes {
+  MULTIPLE = 'multiple',
+  TAGS = 'tags',
+  SINGLE = '',
+}
+export interface SelectProps extends PickedSelectProps {
+  allowNewOptions?: boolean;
+  ariaLabel: string;
+  header?: ReactNode;
+  name?: string; // discourage usage
```
Comment (pkdotson) [inline superset-frontend/src/components/Select/AntdSelect.tsx:69]: I'm curious why this a prop we don't want to use anymore?

Comment (geido) [reply superset-frontend/src/components/Select/AntdSelect.tsx:72]: There's has been a discussion for this prop. The consensus is that `aria-label `is a better option

**Review by pkdotson — APPROVED** [review]
I don't see anything majorly wrong. Looks good to go for the next iteration.

**Review by michael-s-molina — APPROVED** [review]
I think this first iteration is great. Thanks for organizing the work so well.

We don't have any conflicts with the other Select component and the new Select is only being used in the Storybook so we can merge it and continue to iteration 2.

**Comment by rusackas** [comment]
> I think this first iteration is great. Thanks for organizing the work so well.

+💯 This is some awesome foundational work, which is going to put Superset on a much more clean and consistent path with these components. Thank you also for the great pre-coding audit process of all the features and props that our "select party" offered. 

For those who haven't seen it, this new component aims to (eventually) replace all of these:
image

**Pull request merged**